src/lib/pages/resolvePageTree.ts

```diff
@@ -108,35 +114,55 @@ export async function resolvePageTree({
   flattenChildren = false,
   includeDeletedPages = false,
   fullPage,
-  pageNodes,
   tx = prisma
 }: PageTreeResolveInput & OptionalPrismaTransaction): Promise<
   TargetPageTree<PageNodeWithPermissions> | TargetPageTreeWithFlatChildren<PageNodeWithPermissions>
 > {
-  const pageWithSpaceIdOnly = pageNodes
-    ? pageNodes.find((node) => node.id === pageId)
-    : await tx.page.findUnique({
-        where: {
-          id: pageId
-        },
-        select: {
-          spaceId: true
-        }
-      });
+  const pageWithSpaceIdOnly = await tx.page.findUnique({
+    where: {
+      id: pageId
+    },
+    select: {
+      spaceId: true
+    }
+  });
 
   if (!pageWithSpaceIdOnly) {
     throw new PageNotFoundError(pageId);
   }
+  const pagesTreeIds = (await tx.$queryRaw`WITH RECURSIVE parents_cte AS (
+    SELECT id, "parentId"
+    FROM "Page"
+    WHERE id = ${pageId}::UUID
+    
+    UNION ALL
+    
+    SELECT p.id, p."parentId"
+    FROM "Page" p
+    INNER JOIN parents_cte ON p.id = parents_cte."parentId"
+  ), children_cte AS (
+    SELECT id, "parentId"
+    FROM "Page"
+    WHERE id = ${pageId}::UUID
+    
+    UNION ALL
+    
+    SELECT p.id, p."parentId"
+    FROM "Page" p
+    INNER JOIN children_cte ON p."parentId" = children_cte.id
+  )
+  SELECT id FROM parents_cte 
+  UNION 
+  SELECT id FROM children_cte;
+`) as Pick<Page, 'id'>[];
```

Comment:
Neat.. are there any gotchas using Recursive queries? can it go on forever  if there is a circular dependency?

Reply:
Thanks alot for raising this! I added some safeguards in the SQL query, and the reducePagesToPageTree to handle this